### ISSUE TITLE
MeshCentral 0.8.63~0.8.65 . [Features] [Background Re-install Agent in remote control]

### ISSUE THREAD

**Issue by Jacky0400143** [Ylianst/MeshCentral#2797]
Hi, 
Separate issue From #2793 

> [Re-install Agent in remote control] 

Is there a way to reinstall a new agent remotely in the background when the desktop control cannot be operated? For example, the new agent has changed its name, and I want to change the installation location or name remotely,
but The computer cannot be operated physically or remotely, only from the background.
;

**Comment by VINISHVV** [comment]
@Jacky0400143 
You can install the meshagent using command line. 
meshagent.exe -fullinstall

**Comment by Jacky0400143** [comment]
@VINISHVV 
Thank you very much for your Comment.

Here are my steps,Please correct me if I misunderstood.

1. Using MeshCentral Files Transfer New Agent to a direction such like "C:\temp"
2. Using MeshCentral Command line do "meshagent.exe -fullinstall"
3. When New Agent Connected to The Server. It should be two nodes in a same name.
4. Using Device Action to Uninstall Agent remotely Old Agent. (This step should not affect the newly installed agent earlier?)
5. When Sucess Remove Old Agent. Delete Old Node in server.

**Comment by VINISHVV** [comment]
> @VINISHVV
> Thank you very much for your Comment.
> 
> Here are my steps,Please correct me if I misunderstood.
> 
> 1. Using MeshCentral Files Transfer New Agent to a direction such like "C:\temp"
> 2. Using MeshCentral Command line do "meshagent.exe -fullinstall"
> 3. When New Agent Connected to The Server. It should be two nodes in a same name.
> 4. Using Device Action to Uninstall Agent remotely Old Agent. (This step should not affect the newly installed agent earlier?)
> 5. When Sucess Remove Old Agent. Delete Old Node in server.

Its correct.
You can also create batch file to automate the process.

**Comment by Jacky0400143** [comment]
@VINISHVV 

I will Try to , thank you for your help.

If anyone has related needs, instruct above.

**Comment by MailYouLater** [comment]
> Its correct.

Um... I'm not so sure that's right... I also have a strong feeling that this is an X, Y problem, where you have one issue (X), you've decided on a solution (Y), now you're a question about the solution you picked instead of actually asking about the issue that you're actually trying to solve.

---

> 2. Using MeshCentral Command line do "meshagent.exe -fullinstall"

I believe the MeshCentral Agent kills the terminal process when it disconnects, so I'm unsure, but it's possible that running the install this way would stop the install as soon as the existing agent is stopped, leaving the agent in a stopped (or possibly even corrupted) state.

> 3. When New Agent Connected to The Server. It should be two nodes in a same name.

Assuming you didn't manually install the existing agent in a nonstandard way, I'm pretty sure you'd be replacing the existing install with the new one.  Also, assuming the install succeeds, I wouldn't be surprised if it just reconnect as the same node again.

> 4. Using Device Action to Uninstall Agent remotely Old Agent.(This step should not affect the newly installed agent earlier?)

Again, unless you manually installed the existing agent in a nonstandard way, the old agent will have been replaced with the new one, so there will be nothing to uninstall.

> 5. When Sucess Remove Old Agent. Delete Old Node in server.

If it didn't simply reconnect as the same node, but instead appeared as a new node, yes, delete the old now from the server.

---

Anyway...  I get why you want to get this done in the background, but... why do you want to reinstall the agent to begin with?  What are your actually trying to accomplish?

**Comment by Jacky0400143** [comment]
@MailYouLater 
Thank you for comment.

Before asking #2797, I have tried many times with multiple computers. Of course, some succeeded and some failed, and I am not sure if there is a better way or remote re-install Agent may cause some impact, so ...

The result of my current test shows that when the new Agent is not in the same directory, it seems that only a new node with the same name will be added, but the old Agent will not be affected.

However, if the Agent does not change, it seems that the actual reinstallation will indeed cause the same agent to be replaced, which may cause some problems. For example, the old agent disconnected during the replacement process, causing the Comment Line to be interrupted, but this part has not been tested yet.

Because when some agents were previously installed, some settings have not been made for the agent, such as changing the name of the agent file or the installation path. 
some new settings on the server side, I am not sure if I need to reinstall the agent to apply the new settings or not.

**Comment by Jacky0400143** [comment]
Hello , 

Sorry to asking one more question.

If new settings on the server side had been change, did I need to do anything to make all agents belong to same server to apply the new policy and settings ?

"agentupdate"Can deal with this problem?
 I have not set manual update rules, so this part of the default should be executed automatically?

Thank you.

**Comment by VINISHVV** [comment]
@Jacky0400143  There is no such changes required in agent side.

**Comment by VINISHVV** [comment]
> @MailYouLater
> Thank you for comment.
> 
> Before asking #2797, I have tried many times with multiple computers. Of course, some succeeded and some failed, and I am not sure if there is a better way or remote re-install Agent may cause some impact, so ...
> 
> The result of my current test shows that when the new Agent is not in the same directory, it seems that only a new node with the same name will be added, but the old Agent will not be affected.
> 
> However, if the Agent does not change, it seems that the actual reinstallation will indeed cause the same agent to be replaced, which may cause some problems. Just like your Comment said , For example, the old agent disconnected during the replacement process, causing the Comment Line to be interrupted, but this part has not been tested yet.
> 
> Because when some agents were previously installed, some settings have not been made for the agent, such as changing the name of the agent file or the installation path.
> some new settings on the server side, I am not sure if I need to reinstall the agent to apply the new settings or not.

@Jacky0400143 - i hope you are trying to reinstall the agent for  customized meshagents installation.

If its not related to customized agents then please revert with your actual issues

**Comment by Jacky0400143** [comment]
@VINISHVV 

Yes , I was trying to reinstall the agent because of some customized meshagents installation. 
I’m just not sure what exactly needs to be done for this step.And I got a lot of help from you, thank you.

At present, I found that the agent installation path and execution settings . Only those things change need to be reinstalled Agent to deploy, 
About the new server settings, no action is required according to the response received, as long as the agent continues to connect with the server.

If there are other issues, I will open another new issue.

Later I will close this case , 
Thanks for @VINISHVV  @MailYouLater  Your Grant Help.
